### PR TITLE
[CALCITE-6645] Support user-defined function without parentheses when db dialect's allowNiladicParentheses property is false

### PR DESCRIPTION
**Pull request by strongduanmu**
* pass SqlSyntax.FUNCTION_ID to SqlUserDefinedFunction when function has no parameters, and allowNiladicParentheses is false
* add udf test case for function without parentheses

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:286]: why remove the static?

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:306]: here as well

File: core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java:67]: you cannot change public constructors - this is a breaking change for clients.
you have to keep this constructor, and add another one.

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/UdfTest.java:450]: why is this change needed?

File: plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java
Comment (mihaibudiu) [inline plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java:47]: why is this change needed?

File: testkit/src/main/java/org/apache/calcite/test/CalciteAssert.java
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/CalciteAssert.java:208]: what happens here?

File: testkit/src/main/java/org/apache/calcite/util/Smalls.java
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/util/Smalls.java:688]: please add a comment explaining that this function returns the constant string "foo".

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java
Comment (strongduanmu) [reply core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:306]: Since we need to call the non-static `config` field inside the toOp method to determine `config.conformance().allowNiladicParentheses()`, the toOp method cannot be a static method.

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java
Comment (strongduanmu) [reply core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:286]: The operatorTable method will call the toOp method. If the toOp method cannot be a static method, then the operatorTable method also needs to remove static.

Comment (strongduanmu) [reply core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:286]: Besides, I found operatorTable is not called anywhere, should we delete this method?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java
Comment (strongduanmu) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java:67]: Hi @mihaibudiu thank you for your suggestion. I found this public constructor, which is only called in the `CalciteCatalogReader#toOp` method. When we add SqlSyntax to the toOp method, the original constructor that does not contain SqlSyntax will not be called.

In this case, does it make sense to keep a constructor that is not called? In addition, I don't know whether this constructor will be called by the user. If the user can call it, then we should indeed keep the original constructor.

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (strongduanmu) [reply core/src/test/java/org/apache/calcite/test/UdfTest.java:450]: When we use the default conformance, its allowNiladicParentheses method returns false, and functions without parameters should not use parentheses. In order to ensure that the existing test cases can run normally, we need to explicitly set the conformance and ensure that the allowNiladicParentheses of the conformance returns true.

File: testkit/src/main/java/org/apache/calcite/test/CalciteAssert.java
Comment (strongduanmu) [reply testkit/src/main/java/org/apache/calcite/test/CalciteAssert.java:208]: Same as the previous one, we need to set the conformance props to control whether the parameterless function can use parentheses.

File: testkit/src/main/java/org/apache/calcite/util/Smalls.java
Comment (strongduanmu) [reply testkit/src/main/java/org/apache/calcite/util/Smalls.java:688]: I'm very sorry that this naming is confusing for you. This function is to return a constant, and this constant has no meaning. It is just for testing niladic function. It has no meaning just like 0 or -1 returned in other functions.

Comment (strongduanmu) [reply testkit/src/main/java/org/apache/calcite/util/Smalls.java:688]: Do you have any suggestions for better naming for constants used like this?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java:67]: Since Calcite is an open-source library you cannot assume that the only code that calls these functions is in the Calcite tree. There are lots of third party projects using Calcite, and any of them may be calling any of the public functions. You cannot break compatibility unless you change the major version of the project.

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:286]: See my comment below about public methods; you cannot change public method signature without changing the major version of the project. That is called "semantic versioning" https://semver.org/

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/UdfTest.java:450]: perhaps a comment explaining that MYSQL_5 allows nilladic functions without parentheses will help people understand the point. That is not obvious.

File: plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java
Comment (mihaibudiu) [reply plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java:47]: The problem I see is that this is affecting all tests that use this connection provider. Do you really want all of these tests to use MYSQL_5 conformance?

File: testkit/src/main/java/org/apache/calcite/util/Smalls.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/util/Smalls.java:688]: The point I am making is that this is a test for a function with no arguments, and there are two kinds of such functions: non-deterministic, and constant. This is actually a constant function, and it is used for testing the syntax. That's what the comment should say.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java
Comment (strongduanmu) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java:67]: Thanks for your suggestion, I will keep the original constructor and add a new one for passing SqlSyntax.

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (strongduanmu) [reply core/src/test/java/org/apache/calcite/test/UdfTest.java:450]: Thank you @mihaibudiu, I will add comment for this change.

File: plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java
Comment (strongduanmu) [reply plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java:47]: Considering the lex used by the connection provider is MySQL, I think it should use MYSQL_5 conformance here, which makes the config consistent. What do you think?

File: testkit/src/main/java/org/apache/calcite/util/Smalls.java
Comment (strongduanmu) [reply testkit/src/main/java/org/apache/calcite/util/Smalls.java:688]: Thanks for your guidance, I will add comments to improve readability.

File: plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java
Comment (mihaibudiu) [reply plus/src/main/java/org/apache/calcite/chinook/CalciteConnectionProvider.java:47]: Perhaps this would work. I guess the fact that the tests passed shows it's not really a problem.

File: testkit/src/main/java/org/apache/calcite/util/Smalls.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/util/Smalls.java:688]: no need to reply on github for every suggestion - if you plan to do it anyway in your next commit.
Please add a new commit to the PR if you make changes, to make reviewing easier.
You will have to squash the commits before they get merged, though, but after the reviews

**Comment by NobiGo** [comment]
I will review it within a week.

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/UdfTest.java:420]: This looks pretty good.
Are there already negative tests that show errors when the conformance is wrong?

**Review by mihaibudiu — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (strongduanmu) [reply core/src/test/java/org/apache/calcite/test/UdfTest.java:420]: Thanks for your suggestion, I have added some negative tests.

Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/UdfTest.java:420]: In general you should wait to squash the commits until the PR has been approved.
This makes reviewing changes easier.

**Review by mihaibudiu — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/test/UdfTest.java
Comment (strongduanmu) [reply core/src/test/java/org/apache/calcite/test/UdfTest.java:420]: Thank you for your guidance. I will pay attention to this point in my future PRs.

**Comment by mihaibudiu** [comment]
@NobiGo should we wait for your review?

**Review by caicancai — APPROVED** [review]
So cool

**Comment by strongduanmu** [comment]
> @NobiGo should we wait for your review?

Hi @NobiGo, do you have time to review this PR? Thank you very much.

**Comment by strongduanmu** [comment]
@NobiGo Thank you for your feedback, @mihaibudiu Could you please merge this PR? Thanks a lot.

**Pull request merged**
